Copyright (c) 2023 Graphcore Ltd. All rights reserved.

# Dolly 2.0 – The World’s First, Truly Open Instruction-Tuned LLM on IPUs – Inference

|  Domain | Tasks | Model | Datasets | Workflow |   Number of IPUs   | Execution time |
|---------|-------|-------|----------|----------|--------------|--------------|
|   NLP   |  Instruction Fine-tuned Text Generation  | Dolly 2.0 | N/A | Inference | recommended: 16 (minimum 4) |  ???   |

[Dolly 2.0](https://www.databricks.com/blog/2023/04/12/dolly-first-open-commercially-viable-instruction-tuned-llm) is a 12B parameter language model trained and instruction fine-tuned by [Databricks](https://www.databricks.com). By instruction fine-tuning the large language model (LLM), we obtain an LLM better suited for human interactivity. Crucially, Databricks released all code, model weights, and their fine-tuning dataset with an open-source license that permits commercial use. This makes Dolly 2.0 the world's first, truly open-source instruction-tuned LLM. In this notebook, we will show you how to run Dolly 2.0 using Graphcore IPUs on Paperspace with your own prompts.

In this notebook you will:
- Create and configure a Dolly inference pipeline.
- Run Dolly inference on a text prompts to generate answers to user specified questions.

This notebook requires a minimum of 4 IPUs to run. This notebook also supports running on a POD16, resulting in a faster pipeline inference speed.

[![Join our Slack Community](https://img.shields.io/badge/Slack-Join%20Graphcore's%20Community-blue?style=flat-square&logo=slack)](https://www.graphcore.ai/join-community)

## Environment setup

<!-- TODO: add a run on gradient button. -->
The best way to run this demo is on Paperspace Gradient's cloud IPUs because everything is already set up for you.

[![Run on Gradient](../../gradient-badge.svg)]()

To run the demo using other IPU hardware, you need to have the Poplar SDK enabled. Refer to the [Getting Started guide](https://docs.graphcore.ai/en/latest/getting-started.html#getting-started) for your system for details on how to enable the Poplar SDK. Also refer to the [Jupyter Quick Start guide](https://docs.graphcore.ai/projects/jupyter-notebook-quick-start/en/latest/index.html) for how to set up Jupyter to be able to run this notebook on a remote IPU machine.

Run the next cell to install extra requirements for this notebook.

In [ ]:
%pip install -r requirements.txt

In [1]:
import os

number_of_ipus = int(os.getenv("NUM_AVAILABLE_IPU", 16))
number_of_ipus

16

## Dolly inference pipeline

Let's begin by loading the inference config for Dolly. A configuration suitable for your instance will automatically be selected.

In [7]:
from utils.setup import dolly_config_setup

config_name = "dolly_pod4" if number_of_ipus == 4 else "dolly_pod16"
config, *_ = dolly_config_setup("config/inference.yml", "release", config_name)
config

2023-04-28 14:10:10 WARNING: Dropping extra args {'attention_tensor_parallel': 8}
2023-04-28 14:10:10 INFO: Starting. Process id: 117568


DollyConfig(model=ModelConfig(layers=36, hidden_size=5120, sequence_length=2048, precision=<Precision.float16: popxl.dtypes.float16>, seed=42, embedding=ModelConfig.Embedding(vocab_size=50280), attention=ModelConfig.Attention(heads=40, rotary_positional_embeddings_base=10000, rotary_dim=32)), execution=Execution(micro_batch_size=1, data_parallel=1, device_iterations=1, io_tiles=1, available_memory_proportion=(0.4,), tensor_parallel=16, code_load=False))

Next, we want to create our inference pipeline. Here we define the maximum
sequence length. Before executing a model on IPUs it needs to be turned into an
executable format by compiling it. This will happen when the pipeline is
created. All input shapes must be known before compiling, so if the maximum
sequence length is changed, the pipeline will need to be recompiled.

*This cell will take approximately 18 minutes to complete, which includes downloading the model weights.*

In [8]:
import api

# changing these parameters will trigger a recompile
sequence_length = 256  # max 2048

dolly_pipeline = api.DollyPipeline(
    config,
    sequence_length=sequence_length,
)

2023-04-28 14:10:12 INFO: Creating session
2023-04-28 14:10:12 INFO: Starting PopXL IR construction


AssertionError: 40 % 16 != 0

And run the pipeline!

In [ ]:
answer = dolly_pipeline("How many islands are there in Scotland?")

2023-04-27 12:09:49 INFO: Attach to IPUs
2023-04-27 12:09:49 INFO: Start inference
2023-04-27 12:09:49 INFO: Input prompt: How many islands are there in Scotland?
2023-04-27 12:09:49 INFO: Response:
There are 732 islands off the coast of mainland Scotland, with the largest being Rùm (pronounced “roo”) – which is home to over 100,000 sheep. St Kilda is the most remote inhabited island, and is now a nature reserve.


2023-04-27 12:10:15 INFO: Output in 25.25 seconds
2023-04-27 12:10:15 INFO: Throughput: 2.22 t/s


We've just ran Dolly with the default parameters. Thanks to instruction fine-tuning, we can give prompts to Dolly as if it were a chatbot, as instruction fine-tuning helps turn a base language model into one more suited for interactive behaviour with humans.

There are a few sampling parameters we can use to control the behaviour of Dolly:
- `temperature` – Indicates whether you want more creative or more factual outputs. A higher value generates more creative outputs and a lower value generates more factual answers. Typical values fall between `0.0` and `1.0`.
- `top_k` – Indicates that only among the highest `top_k` probable tokens should be sampled. Set to 0 to sample across all possible tokens, which means that top k sampling is disabled. The value for `top_k` must be between a minimum of 0 and a maximum of `config.model.embedding.vocab_size` which is 50,280.
- `output_length` – Indicates the number of tokens to sample before stopping. Sampling can stop early if the model outputs `### END`.
- `print_info` – If `True`, prints extra information such as the total time taken and throughput.
- `print_live` – If `True`, the tokens will be printed as they are being sampled. If  `False`, only the answer will be returned as a string.
- `prompt` – A string containing the question you wish to generate an answer for.

Changing these parameters will not result in any recompilation as the input shape to the model has not changed. These can be freely changed and experimented with in the next cell to produce the behaviour you want from Dolly.

In [9]:
temperature = 0.6
top_k = 5
output_length = None
print_live = True
print_final = True

prompt = "Who was Dolly the sheep?"
answer = dolly_pipeline(
    prompt,
    temperature=temperature,
    k=top_k,
    output_length=output_length,
    print_live=print_live,
    print_final=print_final,
)

NameError: name 'dolly_pipeline' is not defined

As Dolly is an instruction fine-tuned model, it was trained with a specific prompt format. Internally, the pipeline will transform your prompts into the correct format. To see the full prompt, you can view the `last_instruction_prompt` attribute on the pipeline:

In [6]:
print(dolly_pipeline.last_instruction_prompt[0])

NameError: name 'dolly_pipeline' is not defined

Remember to detach your pipeline when you are finished to free up resources:

In [6]:
dolly_pipeline.detach()

## Conclusion

In this notebook, we have demonstrated how you can easily run Dolly 2.0 for inference on Graphcore IPUs on text prompts. Instruction fine-tuning is a powerful method of turning a base LLM into one more suited for human interactivity, such as a question-answer model or a chatbot.

Although larger instruction LLMs exist with more world knowledge such as ChatGPT, they are closed-source or are subject to non-commercial licensing. This makes Dolly 2.0 a significant milestone as the first of its kind, with future instruction fine-tuned LLMs no doubt to quickly follow. All of which will be truly open.

## Next steps

Check out the full list of [IPU-powered Jupyter Notebooks](https://www.graphcore.ai/ipu-jupyter-notebooks) to get more of a feel for how IPUs perform on other tasks.